### PR TITLE
[OoT] New Exporter

### PR DESCRIPTION
**Pull request by Yanis42**
This is something I wanted to do for a while, it should work properly according to my tests (thanks again to Richie for that)

This relies on dataclasses, I don't think it's the best thing for performances but at least it makes the code easy to follow imo, also it includes functions to get C data so it's kinda 2 in 1, this makes you able to call one of the classes to export a specific type of data, I did that for cutscenes and collisions (collisions required more changes but cutscenes are a one-line change)

Anyway this was fun to make!

---
**Any feedback/suggestions is appreciated!** (the diff is huge but this doesn't edit a lot of the original files, most of the diff comes from new files I created)

Extra note: this fix the known issues related to waterbox duplicates, paths and the file(s) being inconsistent at export time, for instance actor entries swapping in the actor list for no reasons

---
This exporter has everything in the same place (inside the ``exporter`` folder) instead of having bits of export scattered here and there, I think it makes more sense like this

### CODE REVIEW

**Comment by Yanis42** [comment]
Lots of updates lately, mainly bugfixes and organisation change/file renames, I'm mostly confident on this working properly, I was able to test most things but I'd like to wait on user feedback

I think I'm happy with how things are organised, I made a (bad) drawing showing the relationship between sub-modules
<details closed>
<summary>File relationships</summary>

</details>

I need to go through the functions and set the definitive names, then I'll probably move on implementing this on other exports and I think it will be good to go after removing the old code (though I'm still using parts of it for textures because I lack knowledge on this)

Also this project fix some bugs with the export, for instance the randomness on actor lists, there's also an issue where spawnpoints aren't in the order defined by entrance spawn indices, all of this is gone!

**Comment by Yanis42** [comment]
todo list:
- [ ] fix conflicts
- [ ] use ``Optional`` for things that can be ``None``
- [ ] complete cutscene stuff
- [ ] remove old code
- [ ] self-review to see if I can improve stuff
- [ ] check everything works
- [ ] implement other exporters (might be part of another pr?)
- [ ] mark as ready to review

**Comment by Yanis42** [comment]
This should be ready to review now, I don't think I'm gonna add something else (it's already huge sorry)
I did almost everything on my todo list, the only remaning item is removing the older code

I decided to keep the checkbox I made to toggle between the two versions, I simply commented the lines, this is on purpose and will be removed with the old version (I can remove that once the review is done and ready to merge of course)

File: fast64_internal/oot/exporter/base.py
Comment (kurethedead) [inline fast64_internal/oot/exporter/base.py:112]: These shouldn't be optional.

**Review by kurethedead — CHANGES_REQUESTED** [review]
Overall I like the organization into more smaller modular parts, but I have a couple issues. Note that I mostly skimmed through the PR, I didn't test it for correctness yet but mostly just for code organization. I didn't mark all of the spots where the following comments apply, just some of them. I'm also a little rusty on OOT stuff so let me know if some of the comments are wrong. 
1. A lot of the classes marked with `@dataclass` shouldn't be dataclasses at all, assuming they're marked as such in order to handle automatic `__init__()` generation. A lot of init-related temporary state is unnecessarily stored as variables, and a lot of derived members are given optional values (where they shouldn't be) in order to "hide" them from being needed in the `__init__()`. The use of `__post_init__()` seems to be a workaround to getting derived state, but at that point its much more clear to just define an `__init__()` function that only takes in the necessary arguments. 
2. A couple classes are just collections of static methods - they should just be functions imported from a separate file, or the methods should be marked `@staticmethod` to avoid class instantiation. For example:
```
from . import spec
...
spec.editSpec(self.exporter)
```

File: fast64_internal/oot/exporter/cutscene/__init__.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/cutscene/__init__.py:24]: I don't like using ``__init__()``, I think when you have args it makes the code not pleasant to read, that's part of the reason why I used dataclasses

I know it's too bad to use ``__post_init__`` but idk I just don't think code like this:
```py
def __init__(arg1, arg2):
    self.arg1 = arg1
    self.arg2 = arg2
```
is aesthetic at all, also using dataclasses makes it obvious which type of class you deal with imo (you could say every class holds data but in this context I mean the data related to the exported scene)

I guess I just misinterpreted the official doc then but I thought it was something there just in case you needed to init more things, not a fan of ``__init__()`` but if it blocks the merge I'll make the change

File: fast64_internal/oot/exporter/main.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/main.py:37]: if this class had less members I'd consider this but I really don't think turning this into a function would be a good thing in this case, it would make a function with 11 parameters (which is too much imo, especially since OoT export is a complex thing), also some of this class members are used somewhere else (to get the scene name, I think it's better this way)

File: fast64_internal/oot/exporter/other/file.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/other/file.py:37]: classes make things more organised imo, I may misuse them a lot in this but at least it's tidy (which isn't a word I use often so I hope this sentence has the meaning I think it does lol)

File: fast64_internal/oot/exporter/room/shape.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/room/shape.py:252]: it's been a while but iirc it's just to keep room shape as a sub-system like everything else, this way if you wanted you could add a button that would only export this, this whole exporter was made based on this idea (which I implemented here for cutscenes and collisions)

**Comment by Dragorn421** [comment]
Fwiw I don't know the pythonic way to have "dataclasses but some of the fields will be initialized later"

I'll try to look it up

File: fast64_internal/oot/exporter/scene/actors.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/scene/actors.py:20]: I know it's weird but I'd like to keep this initialised to ``None`` so if there's any uncaught/unknown issue it would make it more obvious what the issue is, and just using ``Optional`` since I was told ``thing: str = None`` is bad

File: fast64_internal/oot/exporter/base.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/base.py:112]: same as https://github.com/Fast-64/fast64/pull/256#discussion_r1468456826

File: fast64_internal/oot/exporter/room/shape.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/room/shape.py:25]: as long as I can still do ``"{ " + f"{self.type}, {self.amountType}, &{self.entryArrayName}" + " }",`` then yes, but idk how to do that properly (I know for ``int``s there's ``enum.IntEnum`` but not for ``str``s)

**Comment by Yanis42** [comment]
(just in case I forget: https://discord.com/channels/874816709855440926/874819026629574717/1200770808927244499)

File: fast64_internal/oot/exporter/cutscene/__init__.py
Comment (kurethedead) [reply fast64_internal/oot/exporter/cutscene/__init__.py:24]: Dataclasses are definitely useful when your `__init__()` takes in all the data fields as arguments, but a lot of the classes here instead take things like blender objects/properties and then fill out the fields from those objects. Having all the data field arguments be optional just to make the dataclass work doesn't make sense, because someone could technically pass in those optional arguments, but have the result not be what they expect since those arguments get overridden.

The dataclass documentation says you can use `InitVar` for init only variables, but that doesn't really apply here, since none of the optional fields are really optional - you specifically only want them to get their values from other objects you pass in. `__post_init__()` can be used to initialize other derived fields, but you would want to use `field(init=False)` in that case, and even then the documentation itself states that its preferable to just define your own `__init__()` function. This feels a bit like a square peg in a round hole situation.

I don't think classes should be annotated with `@dataclass` solely for documentation purposes, but if you want to keep it, you should define a custom `__init__()` at least. Or, you could create a `@staticmethod` that creates the dataclass object based on only the necessary arguments, although this is basically the same thing as an `__init__()`.

File: fast64_internal/oot/exporter/main.py
Comment (kurethedead) [reply fast64_internal/oot/exporter/main.py:37]: I see, so the class is used both to export and to pass around export parameters. I think it would be better to rename it something like `SceneExport` or `SceneExportParams`, as that's how its used when passing it to Spec or Files. The naming is a little weird because the Spec/Files functions are technically part of the export process, but passing an class named Exporter to it implies that exporting is a sub aspect of those functions. I'm also noticing that the remove scene operator still uses the the old spec functions, not sure if that's intended.

File: fast64_internal/oot/exporter/other/file.py
Comment (kurethedead) [reply fast64_internal/oot/exporter/other/file.py:37]: In that case you should use `@staticmethod` for these functions so you don't have to create an object to use them.

File: fast64_internal/oot/exporter/room/shape.py
Comment (kurethedead) [reply fast64_internal/oot/exporter/room/shape.py:252]: Sorry I'm not understanding, what about that prevents using different subclasses instead of having one class that switches its functionality based on type?

File: fast64_internal/oot/exporter/scene/actors.py
Comment (kurethedead) [reply fast64_internal/oot/exporter/scene/actors.py:20]: Wouldn't it be easier to catch errors at construction time of the `TransitionActor` instead of letting this continue on with invalid state?

File: fast64_internal/oot/exporter/room/shape.py
Comment (kurethedead) [reply fast64_internal/oot/exporter/room/shape.py:25]: You can use `enumValue.name`, but I realize this is probably overkill for this situation so I retract my comment.

**Comment by kurethedead** [comment]
Also, I'm a bit confused about the overall file organization at this point - is the new `exporter` folder going to replace all the scene/room/etc. related exporter files? Is DL/skeleton/animation/collision exporting going to be a part of the `exporter` folder or will it be separate?

**Comment by Yanis42** [comment]
> Also, I'm a bit confused about the overall file organization at this point - is the new `exporter` folder going to replace all the scene/room/etc. related exporter files? Is DL/skeleton/animation/collision exporting going to be a part of the `exporter` folder or will it be separate?

for now they will stay because it would make this PR way too big (and painful for conflicts) but I want to remove everything called when you use ``ootExportSceneToC`` (basically it will remove ``oot_level_writer.py`` and ``scene/exporter/to_c/`` if this gets merged

File: fast64_internal/oot/exporter/cutscene/__init__.py
Comment (Dragorn421) [reply fast64_internal/oot/exporter/cutscene/__init__.py:24]: I'm not sure how to solve this but it sounds to me like the builder pattern (more or less) may be an idea here?

Like you create an object with nothing and iteratively .setThing / .withThing on it (maybe it checks some things don't overwrite previous data too).
Then at the end you .build and it checks everything is there and builds a dataclass with all the things collected

However the builder pattern is not pythonic at all, I've never seen it in python code

Comment (Yanis42) [reply fast64_internal/oot/exporter/cutscene/__init__.py:24]: I looked at it again and I don't see how any of the parameters are overriden by post init, none of the arguments you pass when calling the class are immediately changed, in fact they never get changed, you're using the parameters to set other class members that don't have a default value, which is the purpose of post init according to the [docs](https://docs.python.org/3.10/library/dataclasses.html#post-init-processing):
```py
@dataclass
class C:
    a: float
    b: float
    c: float = field(init=False)

    def __post_init__(self):
        self.c = self.a + self.b
```
this is exactly what I do basically everywhere here, my only mistake to my knowledge is doing ``c: Optional[float] = None`` instead of ``c: float = field(init=False)``, so this is the only change I'll agree on unless there's something I completely misunderstood about dataclasses

Comment (Yanis42) [reply fast64_internal/oot/exporter/cutscene/__init__.py:24]:
```py
@dataclass
class Cutscene:
    """This class defines a cutscene, including its data and its informations"""

    csObj: Object
    useMacros: bool
    name: Optional[str] = None
    motionOnly: bool = False
    data: Optional[CutsceneData] = None
    totalEntries: int = 0
    frameCount: int = 0
    paramNumber: int = 2

    def __post_init__(self):
        # when csObj is None it means we're in import context
        if self.csObj is not None and self.data is None:
            if self.name is None:
                self.name = self.csObj.name.removeprefix("Cutscene.").replace(".", "_")
            self.data = CutsceneData(self.csObj, self.useMacros, self.motionOnly)
            self.totalEntries = self.data.totalEntries
            self.frameCount = self.data.frameCount
```
calling this class looks like this: ``Cutscene(csObj, True)``, I don't see how these are overriden in the post init, unless you meant the other members but that would fall under the mistake I mentioned earlier (which would solve your other concern: https://github.com/Fast-64/fast64/pull/256#discussion_r1468267514 I think)

File: fast64_internal/oot/exporter/room/shape.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/room/shape.py]: I wrote that based on decomp structs, also what I said earlier too to be consistent

also I don't understand that question, what's the concern here exactly? 🤔

File: fast64_internal/oot/exporter/scene/__init__.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/scene/__init__.py]: this should be solved with ``field(init=False)``

File: fast64_internal/oot/exporter/scene/actors.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/scene/actors.py]: I think I can consider this as solved too (``field(init=False)``), the members are set in the same function as the one that calls this class in the first place (which is ``SceneTransitionActors``'s post init)

File: fast64_internal/oot/exporter/other/file.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/other/file.py:37]: I think this PR should wait for #306 and #272 since they both improve the python code and fix issues related to the spec file and the scene table output

File: fast64_internal/oot/exporter/main.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/main.py]: changed it to ``SceneExport``

File: fast64_internal/oot/exporter/room/header.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/room/header.py]: let me know if the new changes are ok to you

File: fast64_internal/oot/exporter/base.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/base.py]: I renamed it "Utility", let me know if you think of a better name (and I used static methods)

**Comment by kurethedead** [comment]
In regards to optional arguments and overriding (https://github.com/Fast-64/fast64/pull/256#discussion_r1484259810), what I meant is that someone could write the code `Cutscene(csObj, true, frameCount = 100)`. However `__post_init__` would immediately override that `frameCount`, which would not only be unintended behaviour - it wouldn't even throw an error at that point, which could be a source of headaches. Using `field(init=False)` does fix this, but doesn't address the other issues I have with the current dataclass usage.

I think there's a discrepancy about how we're looking at dataclasses. From what I'm understanding, dataclasses are supposed to be for classes that are primarily for storing state. Using it for the f3d command classes makes sense in that case, as those only really store state and are operated on by other functions/classes. However, in a lot of cases the export dataclasses have the export functionality embedded in them, which leads to a lot of questionable design decisions.

1. Constructor-specific arguments (including blender objects) being stored as members of the dataclass. I think in almost all cases blender objects shouldn't be members on these classes, as the classes are supposed to be pure data representations of blender objects/properties already. As for non-blender object arguments, an example of confusing design choices is with `CutsceneCmdBase` and its subclasses, particularly with the way `params` and `paramNumber` is handled in order to fit it into the dataclass syntax.
    - `params`, which should only appear in the child class's `__init__`, is now defined as a member in the base class and is queried in all child classes.
    - `paramNumber`, which should be defined in the base class, is separately defined in every child class.

2. Using `field(init=False)` on almost all arguments, then setting all the arguments either externally, or from other constructor arguments. The fact that this has to be done kind of defeats the whole purpose of using dataclasses for `__init__` generation. The former case in particular basically makes all arguments effectively optional, which they shouldn't be.

I think a good middle ground would be a solution like this: https://stackoverflow.com/a/72519184. Basically, the dataclass doesn't use any `field(init=False)`, which indicates that the object should be constructed with all arguments provided and no optional values. Then we have a @class_method that constructs the object based on constructor-only arguments (ex. blender objects), which we no longer have to store as members, instead of using `__post_init__`. This would also fix other design choices made to circumvent dataclass functionality (ex. `paramNumber` can be defined only once in `CutsceneCmdBase`)

**Comment by kurethedead** [comment]
Also, are all the cutscene commands defined twice in different files but with identical definitions?

**Comment by Yanis42** [comment]
> Also, are all the cutscene commands defined twice in different files but with identical definitions?

I wanna do this change in two rounds (or more if needed), round 1 the refactored exporter (this PR), round 2 remove all of the unused code and if that's too long round 3 adapt the remaining cutscene stuff to this, which is the cutscene importer

So yea I know it's not ideal but I'm concerned about PR sizes

**Comment by Yanis42** [comment]
> In regards to optional arguments and overriding ([#256 (comment)](https://github.com/Fast-64/fast64/pull/256#discussion_r1484259810)), what I meant is that someone could write the code `Cutscene(csObj, true, frameCount = 100)`. However `__post_init__` would immediately override that `frameCount`, which would not only be unintended behaviour - it wouldn't even throw an error at that point, which could be a source of headaches. Using `field(init=False)` does fix this, but doesn't address the other issues I have with the current dataclass usage.
> 
> I think there's a discrepancy about how we're looking at dataclasses. From what I'm understanding, dataclasses are supposed to be for classes that are primarily for storing state. Using it for the f3d command classes makes sense in that case, as those only really store state and are operated on by other functions/classes. However, in a lot of cases the export dataclasses have the export functionality embedded in them, which leads to a lot of questionable design decisions.
> 
>     1. Constructor-specific arguments (including blender objects) being stored as members of the dataclass. I think in almost all cases blender objects shouldn't be members on these classes, as the classes are supposed to be pure data representations of blender objects/properties already. As for non-blender object arguments, an example of confusing design choices is with `CutsceneCmdBase` and its subclasses, particularly with the way `params` and `paramNumber` is handled in order to fit it into the dataclass syntax.
>        
>        * `params`, which should only appear in the child class's `__init__`, is now defined as a member in the base class and is queried in all child classes.
>        * `paramNumber`, which should be defined in the base class, is separately defined in every child class.
> 
>     2. Using `field(init=False)` on almost all arguments, then setting all the arguments either externally, or from other constructor arguments. The fact that this has to be done kind of defeats the whole purpose of using dataclasses for `__init__` generation. The former case in particular basically makes all arguments effectively optional, which they shouldn't be.
> 
> 
> I think a good middle ground would be a solution like this: https://stackoverflow.com/a/72519184. Basically, the dataclass doesn't use any `field(init=False)`, which indicates that the object should be constructed with all arguments provided and no optional values. Then we have a @class_method that constructs the object based on constructor-only arguments (ex. blender objects), which we no longer have to store as members, instead of using `__post_init__`. This would also fix other design choices made to circumvent dataclass functionality (ex. `paramNumber` can be defined only once in `CutsceneCmdBase`)

so basically you want me to revert the 1 hour and a half-ish of work I did yesterday to use ``field(init=False)`` (idr how long but it felt long), also I still don't understand why this is a bad thing, I read the [proposal](https://peps.python.org/pep-0557) and the documentation yesterday and my conclusion was dataclasses are meant to be used to make the code prettier, atm everything works exactly as if it wasn't dataclasses so I'm highly confused about why this is bad

it seems to me it's down to what we want in the codebase and not what Python tells us to do, I'm just tired of changing the ~50 different classes this PR adds + testing everything works, fixing issues etc, I usually spent a whole day doing everything and I just wanna move on something else at this point, and I already spent a lot of time working on this (since september, so like 5 months)

btw the whole point of this system is:
1. instanciate a class with the required parameters
2. ready to export

and I feel what you want just breaks this entirely idk, it makes me feel like all of the time I spent on this was basically for nothing lol

I'd like to hear what @Dragorn421 thinks of this and if he got other ideas for a solution before doing anything else, I don't wanna do lots of changes again if it turns out it wasn't necessary

**Comment by kurethedead** [comment]
I'm sorry about the extra work, but I did state that I thought `init=False` was not a good fit for the issues at hand, so I'm not sure why you went ahead with that without more discussion. I think you're missing the main point, which is that there are other issues related to the current code organization besides `init=False` that you aren't addressing. What I'm describing is a code debt issue, not a style issue. The fix I suggested wouldn't break the system you're describing at all, its still the same basic idea.

**Comment by Yanis42** [comment]
I'll try to do the changes soon, I think it's dumb that my own opinions are blocking this cool thing

**Review by kurethedead — CHANGES_REQUESTED** [review]
The changes are good so far - there's just a couple spots missed where factory functions should be used to replace storing temporary objects as members. (Also, some other minor code organization issues). After this PR is merged I think the deletion of the old code should happen ASAP - its pretty confusing when trying to review seeing the same class defined twice in the codebase. For future PRs the deletion should be part of the PR.

File: fast64_internal/oot/exporter/collision/__init__.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/collision/__init__.py]: I guess this comment is old now, this class was renamed to ``CollisionUtility`` and each method is using ``@staticmethod``

the reason why I made it separate was to put away the conversion stuff and only keep the relevant code in ``CollisionHeader``, let me know if I still should make this change though

File: fast64_internal/oot/exporter/collision/surface.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/collision/surface.py]: looks like I got rid of this since, I feel like these comments were made before I done the last review 🤔

File: fast64_internal/oot/exporter/base.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/base.py:108]: (I removed ``@dataclass`` to address this)

File: fast64_internal/oot/exporter/classes.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/classes.py:9]: I went with file.py, I can't find a better name

File: fast64_internal/oot/exporter/collision/__init__.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/collision/__init__.py:233]: meshObj is always set to None is seems, idk how that happened

File: fast64_internal/oot/collision/exporter/to_c/collision.py
Comment (Yanis42) [reply fast64_internal/oot/collision/exporter/to_c/collision.py:300]: (this mistake is the origin of my dumbness above)

File: fast64_internal/oot/exporter/cutscene/__init__.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/cutscene/__init__.py:102]: it doesn't even use the functions apparently

File: fast64_internal/oot/exporter/cutscene/common.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/cutscene/common.py:15]: this is a little bit tricky to do, lots of sub-element of cutscenes actually set these values, sometimes it's just the startFrame, for now I'll just remove the default value

File: fast64_internal/oot/exporter/cutscene/seq.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/cutscene/seq.py:13]: I can't do that everywhere because of ``CutsceneData`` initialising the lists of entries

File: fast64_internal/oot/exporter/other/scene_table.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/other/scene_table.py:131]: I chose to add ``line = line.strip()`` so we remove whitespace everywhere, that's the parsing part so it should be fine

File: fast64_internal/oot/exporter/other/spec.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/other/spec.py:134]: what's a factory function?

File: fast64_internal/oot/exporter/cutscene/common.py
Comment (kurethedead) [reply fast64_internal/oot/exporter/cutscene/common.py]: I don't remember the exact context of my original comment, but I see now that most cutscene commands use startFrame and endFrame in their constructor. I think just removing the default value is good.

Comment (Yanis42) [reply fast64_internal/oot/exporter/cutscene/common.py]: this is already done so I'll consider this as solved

File: fast64_internal/oot/exporter/other/scene_table.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/other/scene_table.py]: remember this? turns out it's breaking the file export lol, I ended up using your suggestion I'm not gonna bother looking deeply into this

File: fast64_internal/oot/exporter/other/scene_table.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/other/scene_table.py]: I was able to remove the first two, but I still need the prefix/suffix because of HackerOoT (and future decomp update™) that have an ifdef to disable debug scenes

File: fast64_internal/oot/exporter/other/scene_table.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/other/scene_table.py]: I managed to remove everything but macro parameters and the prefix/suffix, which are set to ``str()`` by default, let me know if I should make changes about it

File: fast64_internal/oot/exporter/other/scene_table.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/other/scene_table.py]: I think this was addressed, usually I put a 👍 but I forgot on this one lol 

lmk if that's not alright

File: fast64_internal/oot/exporter/other/scene_table.py
Comment (Yanis42) [reply fast64_internal/oot/exporter/other/scene_table.py]: I remember there was a reason but idr what it was

Comment (Yanis42) [reply fast64_internal/oot/exporter/other/scene_table.py]: well, ok it seems to work without that, gonna throw an error instead then

**Review by kurethedead — APPROVED** [review]
I've only done very basic tests, but in terms of code organization I think this is good to go. I'll merge this soon unless any issues appear.

**Pull request merged**
